### ISSUE TITLE
seg fault with gcc 8.1.0

### ISSUE THREAD

**Issue by lee218llnl** [LLNL/LaunchMON#52]
If I `module load gcc/8.1.0` and then build launchmon, I get a seg fault. Here is how I built on LLNL systems:

./configure --prefix=/collab/usr/global/tools/launchmon/toss_3_x86_64_ib/launchmon-test2/ --with-test-rm=slurm --with-test-ncore-per-CN=2 --with-test-nnodes=1 --with-test-rm-launcher=/usr/bin/srun --with-test-installed=install-tests

I then `cd test/src` and run `test_attach.1`. The stack trace looks like this:
```
     _ZN9__gnu_cxx27__verbose_terminate_handlerEv,                                                                                                    FP=7fffffffa3d0
     _ZN10__cxxabiv111__terminateEPFvvE,                                                                                                              FP=7fffffffa3e0
     _ZSt9terminatev,                                                                                                                                 FP=7fffffffa3f0
     __cxa_rethrow,                                                                                                                                   FP=7fffffffa400
C++  std::vector<resource_manager_t,std::allocator<resource_manager_t> >::_M_realloc_insert<const resource_manager_t&>,                               FP=7fffffffa460
C++  std::vector<resource_manager_t,std::allocator<resource_manager_t> >::push_back,                                                                  FP=7fffffffa7a0
C++  rc_rm_t::init,                                                                                                                                   FP=7fffffffa7a0
     LMON_fe_init,                                                                                                                                    FP=7fffffffa800
C++  main,                                                                                                                                            FP=7fffffffbc40
     __libc_start_main,                                                                                                                               FP=7fffffffbd00
     _start,                                                                                                                                          FP=7fffffffbd08
```

I will try to debug this a bit, but let me know if you have any immediate insights.

**Comment by dongahn** [comment]
Seems this errors out at https://github.com/LLNL/LaunchMON/blob/master/launchmon/src/sdbg_rm_map.cxx#L435

**Comment by lee218llnl** [comment]
And the problem goes away if I configure with "--enable-debug"

**Comment by dongahn** [comment]
My guess is under gcc/8.1.0, somehow the `supported_rms` class member vector reallocated this vector and an exception is raised during the copy. Let me see if `resource_manager_t` class is copy construction safe.

**Comment by dongahn** [comment]
> And the problem goes away if I configure with "--enable-debug"

It could be the case that under `--enable-debug` (meaning "-O0 -g"), the initial vector allocation size is much bigger...

**Comment by dongahn** [comment]
So at least `class resource_manager_t` define its copy constructor and assignment operator:

https://github.com/LLNL/LaunchMON/blob/master/launchmon/src/sdbg_rm_map.cxx#L126
https://github.com/LLNL/LaunchMON/blob/master/launchmon/src/sdbg_rm_map.cxx#L152

It looks like they look pretty harmless -- copying basic types like integer, `std::string` and `std::vector<std::string>` etc.

**Comment by dongahn** [comment]
It would be helpful if we know what exception is being thrown here.

**Comment by dongahn** [comment]
@lee218llnl: perhaps you can enclose `init` function with try/catch blocks to find out more about the exception. My guess is this is an "standard" exception derived from `std::exception` class. So could you make the following modifications?

```c++
bool rc_rm_t::init(const std::string &os_isa_string) {
try {
    // init code
} except (std::exception& ex) {
    std::cerr << "Standard exception occurred: " << ex.what() << std::endl;
} except (...) {
    std::cerr << "Unknown exception raised" << std::endl;
}
```

**Comment by lee218llnl** [comment]
looks like it seg faults on unwinding with try catch

**Comment by lee218llnl** [comment]
bash-4.2$ ./test.attach_1
Unknown exception raised


'''
Stack Trace
C    _Unwind_DeleteException,                                                                FP=7fffffffa490
     _ZN7rc_rm_t4initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE.cold.241,       FP=7fffffffa7a0
     LMON_fe_init,                                                                           FP=7fffffffa800
C++  main,                                                                                   FP=7fffffffbc40
     __libc_start_main,                                                                      FP=7fffffffbd00
     _start,                                                                                 FP=7fffffffbd08
'''

**Comment by dongahn** [comment]
> Unknown exception raised

Wow so this isn't a standard exception... There is a chance that this is a memory corruption. 

```c++
rc_rm_t::init(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [clone .cold]
```

**Comment by dongahn** [comment]
Are you in a position to apply Valgrind to the test front end executable?

**Comment by dongahn** [comment]
You may get a lot of reports from that because we don't have a suppression filter. But it may be worth trying.

**Comment by dongahn** [comment]
How about 

```
valgrind  ---tool=memcheck --leak-check=full --trace-children=no --child-silent-after-fork=yes --num-callers=30 --leak-resolution=med --error-exitcode=1 fe_attach_smoketest ...
```

**Comment by dongahn** [comment]
BTW, I have to guess the exception is caught from the `Unknown exception raised` message. If you `exit(1)` in that catch branch, the backtrace won't be generated?

**Comment by dongahn** [comment]
Ah...

[This](https://stackoverflow.com/questions/4885334/c-finding-the-type-of-a-caught-default-exception/24997351#24997351) might also be a good debugging trick in this case. Though it is not "portable" but we are talking about one-off debugging.

**Comment by lee218llnl** [comment]
I ran it under inspector, results attached. 

[inspex-lmon.txt](https://github.com/LLNL/LaunchMON/files/5144587/inspex-lmon.txt)

You can also `xsu lee218` and run inspxe-gui as me and open the results in /g/g0/lee218/src/launchmon/test/src/r001mi3

**Comment by lee218llnl** [comment]
Might be more informative without the try catch:

[inspex-lmon2.txt](https://github.com/LLNL/LaunchMON/files/5144593/inspex-lmon2.txt)

Or /g/g0/lee218/src/launchmon/test/src/r002mi3

**Comment by lee218llnl** [comment]
Looks like it's pointing to an illegal memory access at https://github.com/LLNL/LaunchMON/blob/40174c871965b514a3e980f34e41451ca3b0a0b5/launchmon/src/linux/lmon_api/lmon_fe.cxx#L3005

**Comment by dongahn** [comment]
That's the caller for `bool rc_rm_t::init(const std::string &os_isa_string)`, Something is happening inside that function but because the call stack is showing the inline functions, this doesn't help.

**Comment by dongahn** [comment]
`_M_realloc_insert` is explained in here: https://krshrimali.github.io/How-Vectors-Work-in-C++-Part-1/

As I suspected earlier, I think what's going on is GCC8 does "realloc" for this vector under optimized flag (I don't know why it does this for small vector but nevertheless) and either during copying or destruction needed fro realloc, it gets this exception. 

I looked at the copy constructor and assignment operators which looked okay. I will look at the destructors.

**Comment by dongahn** [comment]
Ugh... silly bug: https://github.com/LLNL/LaunchMON/blob/master/launchmon/src/sdbg_rm_map.cxx#L147


If you change the above line to 

```c++
kill_signals.clear ()
```

Will that avoid double free? Sorry about this.

**Comment by dongahn** [comment]
The other thing is you actually won't need any of those manual free of `std::vector` as this will be handled automatically. Maybe the patch should be to remove the function body of that destructor.

**Comment by dongahn** [comment]
On a second thought, this might not be the cause in and of itself because calling `clear()` twice shouldn't lead to double free.

**Comment by lee218llnl** [comment]
I spent some time debugging this further and this may be an issue with the system boost. When I configure as follows:

```
 CFLAGS="-I/usr/tce/packages/boost/boost-1.72.0-mvapich2-2.3-gcc-8.1.0/include/" CXXFLAGS="-I/usr/tce/packages/boost/boost-1.72.0-mvapich2-2.3-gcc-8.1.0/include/" ./configure --prefix=/collab/usr/global/tools/launchmon/toss_3_x86_64_ib/launchmon-test2/ --with-test-rm=slurm --with-test-ncore-per-CN=2 --with-test-nnodes=1 --with-test-rm-launcher=/usr/bin/srun --with-test-installed=install-tests
```

It runs OK and inspector does not report any illegal memory accesses. I don't have a complete grasp on this, but I had tracked this down to boost::escaped_list_separator, which gave me the idea of trying a newer boost version. I'm not sure if the escaped_list_separator was the direct cause of the seg fault or if perhaps chaning to a newer boost version just masked the real bug. I may poke around this a bit more, but would be happy to assume the former (bug in old boost). Let me know if you have any thoughts and perhaps we can close the issue.

**Comment by lee218llnl** [comment]
boost 1.58 works OK too. The system boost is 1.53.

**Comment by dongahn** [comment]
I wouldn't be surprised if this is a boost issue. Thought BOOST is a respected C++ library, I got bitten by a boost bug in the past as well: https://github.com/flux-framework/flux-sched/issues/366

If you want to detect buggy boost versions (or versions that are known to break us), we should consider using m4 like https://github.com/flux-framework/flux-sched/blob/master/configure.ac#L59 to fail our configure.

Do you think you can add this and a minor fix for https://github.com/LLNL/LaunchMON/issues/52#issuecomment-683209847 and do a quick PR?

**Comment by lee218llnl** [comment]
I'm a bit hesitant about since I haven't definitely confirmed that this is a boost bug and if it is, I don't know where the fix was implemented between 1.53 and 1.58. Also, I"m not sure who will be using such an old version of boost anyway! If I can pin this down, then I will submit a PR, otherwise I don't think this is a big issue, and hopefully we will remember it if someone else reports similar problems.

**Comment by lee218llnl** [comment]
FYI, I'm getting similar seg faults on our rzwhamo system (TOSS 4/RHEL8) which uses gcc-8.3.1 and has boost 1.66.0. I think this will need some further debugging, but the seg fault is in a vector push_back again.

bash-4.4$ head segfault.txt 
Stack Trace
     _Unwind_Resume,                                                                                                          FP=7fffffff15a0
     _ZN18resource_manager_tC2ERKS_.cold.247,                                                                                 FP=7fffffff1600
C++  __gnu_cxx::new_allocator<resource_manager_t>::construct<resource_manager_t,const resource_manager_t&>,                   FP=7fffffff1660
C++  std::allocator_traits<std::allocator<resource_manager_t> >::construct<resource_manager_t,const resource_manager_t&>,     FP=7fffffff1660
C++  std::vector<resource_manager_t,std::allocator<resource_manager_t> >::_M_realloc_insert<const resource_manager_t&>,       FP=7fffffff1660
C++  std::vector<resource_manager_t,std::allocator<resource_manager_t> >::push_back,                                          FP=7fffffff19a0
C++  rc_rm_t::init,                                                                                                           FP=7fffffff19a0
     LMON_fe_init,                                                                                                            FP=7fffffff1a00
C++  STAT_FrontEnd::launchDaemons,                                                                                            FP=7fffffff3b70

The line of code is sdbg_rm_map.cxx line 435:

supported_rms.push_back(a_rm)

**Comment by dongahn** [comment]
Is there any chance a library built with a different version of c++ compilers than the LaunchMON proper (thus slightly different implementation of C++ standard libraries were used? 

I vaguely remember I had a similar issue with `push_back` under such a condition. C++ doesn't have ABI compatibility often even across different libraries each built with different C++ compilers. (or versions)

**Comment by lee218llnl** [comment]
As far as I can tell, they are all built with the system's GNU 8.3.1 compiler. In the absence of more evidence, we're kind of shooting in the dark, so I will try to debug some more.

**Comment by lee218llnl** [comment]
And of course, it doesn't seg fault if I configure with --enable-debug. Thus is the nature of memory bugs...

**Comment by lee218llnl** [comment]
I think I may have a handle on this based on a similar experience I recently had with another user issue. I believe the problem is that in lmon_fe.cxx, you declare a `static rc_rm_t resmanager;`. I think that by doing so, some members (i.e., the vector) may not be allocated properly. If I instead change it to `static rc_rm_t *resmanager;` and make the appropriate changes to call new and change `.` to `->`, then I get past the seg fault. I will have to do some research of the C/C++ standard to see if this is indeed the issue or if perhaps my change is just masking the seg fault. Let me know if you have any insight on this.

**Comment by lee218llnl** [comment]
I appear to have been mistaken in my diagnosis above. I had actually set CXXFLAGS=-g in that build which appears to be what "fixed" the seg fault. Without -g and with my change above, I still get a seg fault. More debugging...

**Comment by lee218llnl** [comment]
I believe the problem was due to = operators that were not returning *this. A fix has been proposed in #53
